webapp/src/main/java/com/box/l10n/mojito/service/pullrun/PullRunAssetService.java

```diff
@@ -59,20 +62,37 @@ public void replaceTextUnitVariants(
             Tags.of("repositoryId", Objects.toString(repository.getId())))
         .record(
             () -> {
-              deleteExistingVariants(pullRunAsset, localeId, outputBcp47Tag);
+              deadLockLoserExceptionRetryTemplate.execute(
+                  context -> {
+                    deleteExistingVariants(pullRunAsset, localeId, outputBcp47Tag);
+                    return null;
+                  });
+
               Lists.partition(uniqueTmTextUnitVariantIds, BATCH_SIZE)
                   .forEach(
                       tuvIdsBatch ->
-                          saveTextUnitVariantsMultiRowBatch(
-                              pullRunAsset, localeId, tuvIdsBatch, outputBcp47Tag));
+                          deadLockLoserExceptionRetryTemplate.execute(
+                              context -> {
+                                saveTextUnitVariantsMultiRowBatch(
+                                    pullRunAsset, localeId, tuvIdsBatch, outputBcp47Tag);
+                                return null;
+                              }));
             });
   }
 
   @Transactional
   void deleteExistingVariants(PullRunAsset pullRunAsset, Long localeId, String outputBcp47Tag) {
     // Delete and insert steps split into two transactions to avoid deadlocks occurring
+
+    // Lock the rows for deletion
+    jdbcTemplate.queryForList(
```

Comment:
it is likely that most "deletes" are no op and that the deadlock comes from the delete statement taking pessimistic locks. 

We should consider checking first if we need to emit the "delete statement" using a select/count. The "for update" should probably be avoided too. 

Just counting the number of rows and optionally emitting the delete can be investigated first.

In case of actual deletes, there might be some interest in fetching the list of ids and passing it down to the delete statement. That would require batching and I'd probably refrain from using the select for update in that case too.

Reply:
Raised #76 containing this approach if we wish to use it instead.